File: solutions_bonustasks/b03_simple_rodstructure/run_LinStaticRod_v10.m
% BK10A6400 Basics of FE Analysis (FEMBasics2023autumn)
% Teacher in charge: Marko Matikainen (MKM)
% Final solution code for the Bonus Task 2.
%
% Goal: The code solves global displacements at node 2 and member forces of a simple rod structure. 
% Coded by MKM for student's usage in the FEMBasics2023 course

%           o N1
%            \
%       Elem1 \   
%              o  N2  
%       Elem2 /|
%            /  F 
%           o N3


% First matlab-based FE code for a simple truss structure
clear all;
close all;
clc;
format shortG;  % just to display values in more convient manner

% Units are mm and N and radians
% Rod 1
E1=210000;
A1=100;
L1=1000;
alpha1=-45*pi/180;
%alpha1=315*pi/180;

% Rod 2
E2=210000;
A2=50;
L2=1000;
alpha2=-135*pi/180;

% Force (in vertical (global) direction at node 2)
F=-100000;


% Transformation into a global coordinate system
% Transformation matrix for a rod 1 
T1=TRod(alpha1);
% Transformation matrix for a rod 2
T2=TRod(alpha2);

% Local elemental stiffness matrices
Kloc1=KlocRod(E1,A1,L1);
Kloc2=KlocRod(E2,A2,L2);

% Global elemental stiffness matrix for a rod 1
Kglob1=T1'*Kloc1*T1;
% Global elemental stiffness matrix for a rod 2
Kglob2=T2'*Kloc2*T2;

% Global stiffness matrix assembling for a whole structure (6x6)
Kglob=zeros(6,6);           % Initialize zero matrix
Kglob(1:4,1:4)=Kglob1;      % Substitute element 1 in a global system stiffness matrix
Kglob(3:6,3:6)=Kglob(3:6,3:6)+Kglob2; % substitute element 2 and sum stiffnesses from elems 1 and 2

% Load vector (in a global coordinate systems)
fglob=zeros(6,1);           % Initialization
fglob(4)=F;                 % Substitute nodal force with respectt to a correct DOF

% Boundary conditions (in a global coordinate systems)
% After bc's, global stiffness matrix is 2x2 and external force vector is
% 2x1.
% Boundary conditions u1,v1,u3,v3 are fixed so corresponding rows and
% columns are eliminated from Kglob and fglob. 
Kglobc=Kglob(3:4,3:4);      % just substitute correct rows and columns
fglobc=fglob(3:4);

% Now it is good to check that the stiffness matrix has full rank and/or det
% ~= 0 to be sure that it is invertible.
rank(Kglobc);
det(Kglobc);

% Solve nodal displacement (in a global coordinate systems)
uglobc=Kglobc\fglobc;

% Let's collect all nodal displacements (solved and fixed)
ugloball=zeros(6,1);
ugloball(3:4)=uglobc;

% Makes elemental nodal displacement vectors
uglob1=ugloball(1:4);
uglob2=ugloball(3:6);

% Solve element's axial displacements (in a local coordinate systems)
uloc1=T1*uglob1;
uloc2=T2*uglob2;

% Let's solve member forces f=K u
floc1=Kloc1*uloc1;
floc2=Kloc2*uloc2;

% Next ones are asked to return in Moodle
disp('Determinant of Kglob')
det(Kglob)             % Determinant of unconstrained global stiff matrix (should be 0 or really close)
disp('Determinant of Kglobc')
det(Kglobc)            % Determinant of constrained global stiff matrix (>>0)
disp('Displacement at node 2')
ugloball(3:4)          % Displacements at node 2
disp('Local elemental displacements')
uloc1
uloc2
disp('Member forces')
floc1
floc2

File: solutions_old_bonustasks/Simple_rod_structure_2rods/python/run_LinStaticRod_v10.py
# BK10A6400 Basics of FE Analysis
# Final solution code for Bonus Task 2
#
# Goal:
# Solve global displacements at node 2
# and member forces of a simple rod structure.
#
# Units: mm, N, radians


import numpy as np

from TRod import TRod
from KlocRod import KlocRod


# ==========================================================
# Rod 1
# ==========================================================

E1 = 210000
A1 = 100
L1 = 1000

alpha1 = -45 * np.pi / 180


# ==========================================================
# Rod 2
# ==========================================================

E2 = 210000
A2 = 50
L2 = 1000

alpha2 = -135 * np.pi / 180


# ==========================================================
# Force
# Force in vertical global direction at node 2
# ==========================================================

F = -100000


# ==========================================================
# Transformation into a global coordinate system
# ==========================================================

# Transformation matrix for rod 1
T1 = TRod(alpha1)

# Transformation matrix for rod 2
T2 = TRod(alpha2)


# ==========================================================
# Local elemental stiffness matrices
# ==========================================================

Kloc1 = KlocRod(E1, A1, L1)
Kloc2 = KlocRod(E2, A2, L2)


# ==========================================================
# Global elemental stiffness matrix for rod 1
# ==========================================================

Kglob1 = T1.T @ Kloc1 @ T1


# ==========================================================
# Global elemental stiffness matrix for rod 2
# ==========================================================

Kglob2 = T2.T @ Kloc2 @ T2


# ==========================================================
# Global stiffness matrix for whole structure
# 6 x 6
# ==========================================================

Kglob = np.zeros((6, 6))


# Substitute element 1 into global stiffness matrix
Kglob[0:4, 0:4] = Kglob1


# Substitute element 2 and add stiffnesses
Kglob[2:6, 2:6] += Kglob2


# ==========================================================
# Load vector in global coordinate system
# ==========================================================

fglob = np.zeros((6, 1))

# Force at DOF 4
fglob[3, 0] = F


# ==========================================================
# Boundary conditions
#
# u1, v1, u3, v3 are fixed
#
# Remaining DOFs:
# u2, v2
# ==========================================================

Kglobc = Kglob[2:4, 2:4]

fglobc = fglob[2:4]


# ==========================================================
# Check rank and determinant
# ==========================================================

print("Rank of Kglobc:")
print(np.linalg.matrix_rank(Kglobc))

print("\nDeterminant of Kglobc:")
print(np.linalg.det(Kglobc))


# ==========================================================
# Solve nodal displacement
# ==========================================================

uglobc = np.linalg.solve(Kglobc, fglobc)


# ==========================================================
# Collect all nodal displacements
# Fixed displacements are zero
# ==========================================================

ugloball = np.zeros((6, 1))

ugloball[2:4] = uglobc


# ==========================================================
# Elemental nodal displacement vectors
# ==========================================================

uglob1 = ugloball[0:4]

uglob2 = ugloball[2:6]


# ==========================================================
# Solve elemental axial displacements
# in local coordinate system
# ==========================================================

uloc1 = T1 @ uglob1

uloc2 = T2 @ uglob2


# ==========================================================
# Solve member forces
# f = K u
# ==========================================================

floc1 = Kloc1 @ uloc1

floc2 = Kloc2 @ uloc2


# ==========================================================
# Display results
# ==========================================================

print("\n========================================")
print("Determinant of Kglob")
print("========================================")

# Determinant of unconstrained global stiffness matrix
# Should be zero or very close to zero

print(np.linalg.det(Kglob))


print("\n========================================")
print("Determinant of Kglobc")
print("========================================")

# Determinant of constrained global stiffness matrix
# Should be > 0

print(np.linalg.det(Kglobc))


print("\n========================================")
print("Displacement at node 2")
print("========================================")

print(ugloball[2:4])


print("\n========================================")
print("Local elemental displacements")
print("========================================")

print("Element 1:")
print(uloc1)

print("\nElement 2:")
print(uloc2)


print("\n========================================")
print("Member forces")
print("========================================")

print("Element 1:")
print(floc1)

print("\nElement 2:")
print(floc2)